### PR TITLE
ISSUE-16731: "msvc" compatibility with "Visual Studio" is broken

### PR DESCRIPTION
**Pull request by harsszegi**
In case cppstd is set in the Conan profile for "Visual Studio", it is badly migrated to "msvc". This change solves: https://github.com/conan-io/conan/issues/16731

Changelog: Bugfix: Fixed msvc / Visual Studio package compatibility in case cppstd is used in profiles
Docs: Omit

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

File: conans/model/info.py [hunk 1]
```diff
@@ -591,8 +591,12 @@ def msvc_compatible(self):
         version = compatible.settings.compiler.version
         runtime = compatible.settings.compiler.runtime
         runtime_type = compatible.settings.compiler.runtime_type
+        cppstd = compatible.settings.compiler.cppstd
```
Comment (memsharded) [inline conans/model/info.py]: There are packages that ``self.settings.rm_safe("compiler.cppstd")``, so this will crash in those cases

**Review by memsharded — COMMENTED** [review]
Many thanks for your contribution @harsszegi 

It would be necessary to add a couple of checks or error protections for cases where ``cppstd`` might not be defined.

It would also be necessary to fix the currently breaking test ``conans/test/integration/package_id/compatible_test.py::test_msvc_visual_incompatible`` and most likely to add some new test that cover the use case. Please ask for guidance if you need it.

Finally, please have a look and sign the CLA, to be able to accept the changes.

File: conans/model/info.py (hunk 1 above)
Comment (harsszegi) [reply conans/model/info.py]: Changed to "get_safe" and check for not None

**Comment by memsharded** [comment]
The CI is still failing with:
```
E           ------------------------ Command failed (unexpectedly): ------------------------
E           install pkg/0.1@ -pr=profile
E           ----------------------------------- Output: ------------------------------------
E           
E           WARN: **************************************************
E           WARN: *** Conan 1 is legacy and on a deprecation path **
E           WARN: *********** Please upgrade to Conan 2 ************
E           WARN: **************************************************
E           Configuration:
E           [settings]
E           arch=x86_64
E           build_type=Release
E           compiler=msvc
E           compiler.cppstd=14
E           compiler.runtime=dynamic
E           compiler.runtime_type=Release
E           compiler.version=191
E           os=Windows
E           [options]
E           [build_requires]
E           [env]
E           
E           Installing package: pkg/0.1
E           Requirements
E               pkg/0.1 from local cache - Cache
E           Packages
E               pkg/0.1:16041e00ee7792b16701a7a92918452f6b5a519b - Missing
E           
E           Cross-build from 'Linux:x86_64' to 'Windows:x86_64'
E           Installing (downloading, building) binaries...
E           ERROR: Missing binary: pkg/0.1:16041e00ee7792b16701a7a92918452f6b5a519b
E           
E           pkg/0.1: WARN: Can't find a 'pkg/0.1' package for the specified settings, options and dependencies:
E           - Settings: arch=x86_64, build_type=Release, compiler=msvc, compiler.cppstd=14, compiler.runtime=dynamic, compiler.runtime_type=Release, compiler.version=191, os=Windows
E           - Options: 
E           - Dependencies: 
E           - Requirements: 
E           - Package ID: 16041e00ee7792b16701a7a92918452f6b5a519b
E           
E           ERROR: Missing prebuilt package for 'pkg/0.1'
E           Use 'conan search pkg/0.1@ --table=table.html -r=remote' and open the table.html file to see available packages
E           Or try to build locally from sources with '--build=pkg'
E           
E           More Info at 'https://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package'
```

in ``test_msvc_visual_incompatible`` test

**Comment by harsszegi** [comment]
@memsharded I'm completely lost on this :(

If I do:

conan create . pkg/0.1@ -s os=Windows -s compiler="Visual Studio" -s compiler.version=15 -s compiler.cppstd=**14** -s compiler.runtime=MD -s build_type=Release -s arch=x86_64

then a 

conan search pkg/0.1@ --table=table.html

cppstd is not set

however if I do a :

conan create . pkg/0.1@ -s os=Windows -s compiler="Visual Studio" -s compiler.version=15 -s compiler.cppstd=**17** -s compiler.runtime=MD -s build_type=Release -s arch=x86_64

then a

conan search pkg/0.1@ --table=table.html

cppstd becomes set.

This is more than problematic as I have zero idea in what circumstances will cppstd be erased :( Could you explain what the heck is going on here and more important why?

Meh, it appears to be the "cppstd_default" hacking to erase cppstd if it's the default :(

Checking how I can detect this situation in the conversion function.

**Comment by memsharded** [comment]
This is due to some automatic behavior in Conan 1: https://docs.conan.io/1/reference/conanfile/methods.html#self-info-default-std-matching-self-info-default-std-non-matching

It is basically making ``cppstd=None`` when it matches the default ``cppstd`` for the current compiler version. It was a way to allow the introduction of ``cppstd`` without breaking.

**Comment by harsszegi** [comment]
Ok, added this default-handling thingie to the conversion function (e.g. it erases cppstd if it matches the 'default')

**Review by memsharded — APPROVED** [review]
Looks good, thanks for your contribution.

Could you please check and sign the CLA, so it can be merged? Thanks!

**Review by AbrilRBS — APPROVED** [review]

**Comment by harsszegi** [comment]
> Looks good, thanks for your contribution.
> 
> Could you please check and sign the CLA, so it can be merged? Thanks!

The sad part is that I have signed it 3x, but still says pending :( (my Github CLA page also lists conan-io there)

**Comment by memsharded** [comment]
> The sad part is that I have signed it 3x, but still says pending :( (my Github CLA page also lists conan-io there)

This might happen when the commits are signed with a different email address than the one used by Github, can you please check that?

**Comment by harsszegi** [comment]
> > The sad part is that I have signed it 3x, but still says pending :( (my Github CLA page also lists conan-io there)
> 
> This might happen when the commits are signed with a different email address than the one used by Github, can you please check that?

F@ck, thanks, I pushed with my corporate Github account instead of my private one :(

**Pull request merged**